source/module_rdmft/rdmft_tools.h

```diff
@@ -0,0 +1,460 @@
+#include "module_basis/module_ao/parallel_orbitals.h"
+#include "module_hamilt_pw/hamilt_pwdft/global.h"
+#include "module_base/parallel_reduce.h"
+#include "module_elecstate/module_dm/cal_dm_psi.h"
+#include "module_elecstate/module_dm/density_matrix.h"
+
+#include "module_hamilt_general/operator.h"
+#include "module_hamilt_lcao/module_hcontainer/hcontainer.h"
+#include "module_hamilt_lcao/hamilt_lcaodft/operator_lcao/operator_lcao.h"
+#include "module_hamilt_lcao/hamilt_lcaodft/operator_lcao/op_exx_lcao.h"
+#include "module_hamilt_lcao/hamilt_lcaodft/operator_lcao/ekinetic_new.h"
+#include "module_hamilt_lcao/hamilt_lcaodft/operator_lcao/nonlocal_new.h"
+#include "module_hamilt_lcao/hamilt_lcaodft/operator_lcao/veff_lcao.h"
+
+#include "module_hamilt_lcao/hamilt_lcaodft/hs_matrix_k.hpp"
+
+// used by Exx&LRI
+#ifdef __EXX
+#include "module_ri/RI_2D_Comm.h"
+#include "module_ri/Exx_LRI.h"
+#endif
+
+// there are some operator reload to print data in different formats
+#include "module_ri/test_code/test_function.h"
+
+#include <iostream>
+#include <type_traits>
+#include <complex>
+#include <vector>
+#include <iomanip>
+
+
+
+namespace rdmft
+{
+
+
+//for print matrix
+template <typename TK>
+void printMatrix_pointer(int M, int N, const TK* matrixA, std::string nameA)
+{
+    std::cout << "\n" << nameA << ": \n";
+    for(int i=0; i<M; ++i)
+    {
+        for(int j=0; j<N; ++j)
+        {
+            if( j%5 == 0 ) { std::cout << "\n";
+}
+            std::cout << *(matrixA+i*N+j) << " ";
+        }
+        std::cout << "\n";
+    }
+    std::cout << std::endl;
+}
+
+
+template <typename TK>
+void printMatrix_vector(int M, int N, const std::vector<TK>& matrixA, std::string nameA)
+{
+    std::cout << "\n" << nameA << ": \n";
+    for(int i=0; i<M; ++i)
+    {
+        for(int j=0; j<N; ++j)
+        {
+            if( j%5 == 0 ) { std::cout << "\n";
+}
+            std::cout << matrixA[i*N+j] << " ";
+        }
+        std::cout << "\n\n";
+    }
+    std::cout << std::endl;
+}
+
+
+// now support XC_func_rdmft = "hf", "muller", "power", "pbe", "pbe0". "wp22" and "cwp22" is realizing.
+// for the dft-xc-functional part of xc-functional, just use the default is right! Or don't use the function
+double occNum_func(double eta, int symbol = 0, const std::string XC_func_rdmft = "hf", const double alpha_power = 1.0);
+
+
+template <typename TK>
+void set_zero_vector(std::vector<TK>& HK)
+{
+    for(int i=0; i<HK.size(); ++i) { HK[i] = 0.0;
+}
+}
+
+
+template <typename TK>
+void set_zero_psi(psi::Psi<TK>& wfc)
+{
+    TK* pwfc_in = &wfc(0, 0, 0);
+
+    #ifdef _OPENMP
+    #pragma omp parallel for schedule(static, 1024)
+    #endif
+    
+    for(int i=0; i<wfc.size(); ++i) pwfc_in[i] = 0.0;
+}
+
+
+template <typename TK>
+void conj_psi(psi::Psi<TK>& wfc)
+{
+    TK* pwfc = &wfc(0, 0, 0);
+    for(int i=0; i<wfc.size(); ++i) { pwfc[i] = std::conj( pwfc[i] );
+}
+}
+
+
+template <>
+void conj_psi<double>(psi::Psi<double>& wfc);
```

Comment:
It's not a good idea to 'invent' several new operations on Psi, these are common tools, not the tools by RDMFT itself.

Reply:
I second this.